pkg/services/sqlstore/migrations/migrations_test.go

```diff
@@ -53,6 +57,126 @@ func TestMigrations(t *testing.T) {
 	checkStepsAndDatabaseMatch(t, mg, expectedMigrations)
 }
 
+func TestMigrationLock(t *testing.T) {
```

Comment:
Note that the tests in this file are not marked as integration tests so won't run during CI against MySQL and Postgres. I'm not sure if we want to force then run during CI checks.

for running them locally:
```
make devenv sources=mysql_tests,postgres_tests mysql_version=5.7
GRAFANA_TEST_DB=mysql go test -v -count 1 --race ./pkg/services/sqlstore/migrations/...
GRAFANA_TEST_DB=postgres go test -v -count 1 --race ./pkg/services/sqlstore/migrations/...
```

Reply:
You think/know it will lock the database under test and affect other tests? Think each integration test runs per package synchronously so should be fine I think. Would be good to run this in CI if possible.

Reply:
Or alternatively, implementing an integration test [here](https://github.com/grafana/grafana/blob/main/pkg/tests/README.md) that starts two instances at the same time perhaps? Not sure how easy that would be. We're not running them with integration build tag today, but we probably want that and you could definitely do that for this test I believe.

Reply:
> You think/know it will lock the database under test and affect other tests? Think each integration test runs per package synchronously so should be fine I think. Would be good to run this in CI if possible.

good question! I unlock the database during test cleanups and also locking should not last long in the tests (don't do anything meaningful) but I believe it is still possible if tests run concurrently. I agree that I would like to include them in the CI.

> Or alternatively, implementing an integration test here that starts two instances at the same time perhaps? Not sure how easy that would be. We're not running them with integration build tag today, but we probably want that and you could definitely do that for this test I believe.

I tried to make an integration test but since `testinfra.StartGrafana` performs also other tasks it was difficult to synchronise them so that the migrations occur at the same time.

Reply:
> I tried to make an integration test but since `testinfra.StartGrafana` performs also other tasks it was difficult to synchronise them so that the migrations occur at the same time.

Okay. So maybe easier with a database integration test then and run two go routines somehow syncing them to race/compete for the database lock.

Reply:
I've added [TestDatabaseLocking](https://github.com/grafana/grafana/blob/d422f312cda0ca480fe4433e037a9dd3bc458eeb/pkg/services/sqlstore/migrations/migrations_test.go#L185) where I create two migrators (for the same database) and run the migrations concurrency and check that the 2nd one fails with the error returned by `Lock()`.
I enclose the test output:
<details>
<summary></summary>

```
GRAFANA_TEST_DB=mysql go test -v -count 1 --race ./pkg/services/sqlstore/migrations/... -run TestDatabaseLocking 
=== RUN   TestDatabaseLocking
=== RUN   TestDatabaseLocking/when_concurrent_migrations_occur_for_different_migrators_occur,_the_second_one_should_fail
=== RUN   TestDatabaseLocking/when_concurrent_migrations_occur_for_different_migrators_occur,_the_second_one_should_fail/run_migration_0
=== PAUSE TestDatabaseLocking/when_concurrent_migrations_occur_for_different_migrators_occur,_the_second_one_should_fail/run_migration_0
=== RUN   TestDatabaseLocking/when_concurrent_migrations_occur_for_different_migrators_occur,_the_second_one_should_fail/run_migration_1
=== PAUSE TestDatabaseLocking/when_concurrent_migrations_occur_for_different_migrators_occur,_the_second_one_should_fail/run_migration_1
=== CONT  TestDatabaseLocking/when_concurrent_migrations_occur_for_different_migrators_occur,_the_second_one_should_fail/run_migration_1
=== CONT  TestDatabaseLocking/when_concurrent_migrations_occur_for_different_migrators_occur,_the_second_one_should_fail/run_migration_0
logger=migrator t=2022-01-31T18:11:43.63+0200 lvl=info msg="Locking database"
logger=migrator t=2022-01-31T18:11:43.63+0200 lvl=info msg="Starting DB migrations"
logger=migrator t=2022-01-31T18:11:43.63+0200 lvl=info msg="Locking database"
logger=migrator t=2022-01-31T18:11:43.64+0200 lvl=eror msg="Failed to lock database" error="failed to obtain lock"
logger=migrator t=2022-01-31T18:11:43.64+0200 lvl=info msg="Executing migration" id="create migration_log table"
[...]
logger=migrator t=2022-01-31T18:11:47.86+0200 lvl=info msg="Unlocking database"
--- PASS: TestDatabaseLocking (4.96s)
    --- PASS: TestDatabaseLocking/when_concurrent_migrations_occur_for_different_migrators_occur,_the_second_one_should_fail (0.00s)
        --- PASS: TestDatabaseLocking/when_concurrent_migrations_occur_for_different_migrators_occur,_the_second_one_should_fail/run_migration_0 (0.01s)
        --- PASS: TestDatabaseLocking/when_concurrent_migrations_occur_for_different_migrators_occur,_the_second_one_should_fail/run_migration_1 (4.24s)
PASS
```
</details>